### PR TITLE
UFP128

### LINKED ISSUE #38: UFP128
### Motivation

Addition of 128-bit fixed-point type - `UFP128`

### PR DESCRIPTION
**Pull request by rostyslavtyshko**
## Type of change

- New feature

## Changes

The following changes have been made:

Addition of 128-bit fixed-point typs - `UFP128`

## Notes

- Note 1

## Related Issues

Closes #38

### CODE REVIEW

**Comment by simonr0204** [comment]
Just for posterity, could you fill out the PR description with the change?

**Comment by bitzoic** [comment]
@rostyslavtyshko Are you able to apply the same comments made on the `UFP64` library PR onto this PR regarding the comments, documentation, and testing?

**Comment by rostyslavtyshko** [comment]
> @rostyslavtyshko Are you able to apply the same comments made on the `UFP64` library PR onto this PR regarding the comments, documentation, and testing?

doing it

File: sway_libs/src/fixed_point/ufp/ufp128/ufp128.sw [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+    fn exp(exponent: Self) -> Self {
+        let one = UFP128::from((1, 0));
+        let p2 = one / UFP128::from((2, 0));
+        let p3 = one / UFP128::from((6, 0));
+        let p4 = one / UFP128::from((24, 0));
+        let p5 = one / UFP128::from((120, 0));
+        let p6 = one / UFP128::from((720, 0));
+        let p7 = one / UFP128::from((5040, 0));
+
+        // common technique to counter loosing sugnifucant numbers in usual approximation
```
Comment (simonr0204) [inline sway_libs/src/fixed_point/ufp/ufp128/ufp128.sw:228]:
```suggestion
        // common technique to counter losing significant numbers in usual approximation
```

File: sway_libs/src/fixed_point/ufp/ufp128/ufp128.sw [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+        assert(!(self.value < other.value));
+
+        UFP128 {
+            value: self.value - other.value,
+        }
+    }
+}
+
+impl core::ops::Multiply for UFP128 {
+    /// Multiply a UFP128 with a UFP128. Panics of overflow.
```
Comment (simonr0204) [inline sway_libs/src/fixed_point/ufp/ufp128/ufp128.sw:99]:
```suggestion
    /// Multiply a UFP128 with a UFP128. Panics on overflow.
```

File: sway_libs/src/fixed_point/ufp/ufp128/ufp128.sw [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+            value: self.value + other.value,
+        }
+    }
+}
+
+impl core::ops::Subtract for UFP128 {
+    /// Subtract a UFP128 from a UFP128. Panics of overflow.
+    fn subtract(self, other: Self) -> Self {
+        // If trying to subtract a larger number, panic.
+        assert(!(self.value < other.value));
```
Comment (simonr0204) [inline sway_libs/src/fixed_point/ufp/ufp128/ufp128.sw:90]:
```suggestion
        assert((self.value >= other.value));
```

File: sway_libs/src/fixed_point/ufp/ufp128/ufp128.sw [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+
+    pub fn fract(self) -> Self {
+        Self::from((0, self.value.lower))
+    }
+}
+
+impl Root for UFP128 {
+    fn sqrt(self) -> Self {
+        let nominator_root = self.value.sqrt();
+        let nominator = nominator_root * U128::from((0, 2 << 32));
```
Comment (simonr0204) [inline sway_libs/src/fixed_point/ufp/ufp128/ufp128.sw:187]: Is `nominator` standard terminology? It sounds like a confusion of "numerator" and "denominator" but maybe this is a fixed-point math term I'm unfamiliar with.

File: sway_libs/src/fixed_point/ufp/ufp128/ufp128.sw [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+        let u128_1 = U128::from((0, 1));
+        let u128_2 = U128::from((0, 2));
+        let u128_64 = U128::from((0, 64));
+        let two_pow_64_n_minus_1 = u128_2.pow(u128_64 * (exponent.value - u128_1));
+        let nominator = nominator_pow / two_pow_64_n_minus_1;
+        Self::from((nominator.upper, nominator.lower))
+    }
+}
+
+// TODO: uncomment and change accordingly, when signed integers will be added
```
Comment (simonr0204) [inline sway_libs/src/fixed_point/ufp/ufp128/ufp128.sw:204]: Signed integers are good to go, no?

**Review by simonr0204 — CHANGES_REQUESTED** [review]

File: sway_libs/src/fixed_point/ufp/ufp128/ufp128.sw (hunk 5 above)
Comment (rostyslavtyshko) [reply sway_libs/src/fixed_point/ufp/ufp128/ufp128.sw:204]: log function should changed, so this part has to stay as is, for now

File: sway_libs/src/fixed_point/ufp/ufp128/ufp128.sw (hunk 3 above)
Comment (rostyslavtyshko) [reply sway_libs/src/fixed_point/ufp/ufp128/ufp128.sw:90]: No method named "ge" found for type "U128", so it won't work. I'll add it to sway, but as this is a minor issue, I'd like this part to be merged as is for now.

File: sway_libs/src/fixed_point/ufp/ufp128/ufp128.sw (hunk 1 above)
Comment (rostyslavtyshko) [reply sway_libs/src/fixed_point/ufp/ufp128/ufp128.sw]: Fixed accordingly

File: sway_libs/src/fixed_point/ufp/ufp128/ufp128.sw (hunk 4 above)
Comment (rostyslavtyshko) [reply sway_libs/src/fixed_point/ufp/ufp128/ufp128.sw]: Fixed accordingly

File: sway_libs/src/fixed_point/ufp/ufp128/ufp128.sw (hunk 2 above)
Comment (rostyslavtyshko) [reply sway_libs/src/fixed_point/ufp/ufp128/ufp128.sw]: Fixed accordingly

**Review by simonr0204 — DISMISSED** [review]

**Comment by simonr0204** [comment]
LGTM, will allow the others time to review (after the holidays) before merging.

**Review by bitzoic — CHANGES_REQUESTED** [review]
A few comments but looks good! These changes shouldn't take long. 

Compared to `ufp64`, there's a lot of inline documentation that `ufp128` does not have. Could you carry these over?

I also assume we will be reusing the same README.md and SPECIFICATION.md from the `ufp64` library. In that case, we need to do a few things:

1. Make sure to remove '64-bit' whenever there is mention of the "Unsigned Fixed Point 64-bit library" as this applies to both 64-bit and 128-bit now. This should be done in the source README.md, the library's README.md, and the library's SPECIFICATION.md.

2. Modify the file structure as so:
sway-libs/
├──  fixed_point_integers/
|          ├──  ufp64.sw
|          ├──  ufp128.sw
|          ├──  README.md
|          └── SPECIFICATION.md
(don't forget to update the link in the main README.md)

3. Include the following missing functions from the SPECIFICATION.md:
 - from()
 - into()
 - denominator()
 - zero()
 - from_uint()
 - floor()
 - fract()
 - ceil()

File: sway_libs/src/fixed_point/ufp/ufp128/ufp128.sw [hunk 6]
```diff
@@ -0,0 +1,234 @@
+library ufp128;
+//! A wrapper around U128 type for a library for Sway for mathematical functions operating with signed 64.64-bit fixed point numbers.
```
[the remark this answers is not in the document]
Comment (rostyslavtyshko) [reply sway_libs/src/fixed_point/ufp/ufp128/ufp128.sw]: Fixed

File: sway_libs/src/lib.sw [hunk 7]
```diff
@@ -11,3 +11,5 @@ dep signed_integers/i128;
 dep signed_integers/i256;
 dep nft/nft;
 dep string/string;
+
+dep fixed_point/ufp/ufp128/ufp128;
```
[the remark this answers is not in the document]
Comment (rostyslavtyshko) [reply sway_libs/src/lib.sw]: Done

File: sway_libs/src/fixed_point/ufp/ufp128/ufp128.sw [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+// TODO: uncomment and change accordingly, when signed integers will be added
+// impl Logarithm for UFP128 {
+//     fn log(self, base: Self) -> Self {
+//         let nominator_log = self.value.log(base);
+//         let res = (nominator_log - U128::from(0, 64 * 2.log(base))) * U128::from(1, 0);
+//         UFP128 {
+//             value: res
+//         }
+//     }
+// }
```
[the remark this answers is not in the document]
Comment (rostyslavtyshko) [reply sway_libs/src/fixed_point/ufp/ufp128/ufp128.sw]: Fixed

File: sway_libs/src/fixed_point/ufp/ufp128/ufp128.sw [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+    }
+
+    /// The size of this type in bits.
+    pub fn bits() -> u32 {
+        128
+    }
+}
+
+impl UFP128 {
+    /// Creates UFP64 that correponds to a unsigned integer
```
[the remark this answers is not in the document]
Comment (rostyslavtyshko) [reply sway_libs/src/fixed_point/ufp/ufp128/ufp128.sw]: Fixed

File: sway_libs/src/fixed_point/ufp/ufp128/ufp128.sw [hunk 10]
```diff
@@ -0,0 +1,234 @@
+library ufp128;
+//! A wrapper around U128 type for a library for Sway for mathematical functions operating with signed 64.64-bit fixed point numbers. 
+use std::u128::U128;
+use std::u256::U256;
+use std::assert::assert;
+use std::math::{Exponent, Exponentiate, Root};
+use std::revert::revert;
```
[the remark this answers is not in the document]
Comment (rostyslavtyshko) [reply sway_libs/src/fixed_point/ufp/ufp128/ufp128.sw]: Fixed

File: README.md [hunk 11]
```diff
@@ -35,7 +35,7 @@ These libraries contain helper functions, generalized standards, and other tools
 - [Non-Fungible Token (NFT)](./sway_libs/src/nft/) is a token library which provides unqiue collectibles, identified and differentiated by token IDs.
 - [String](./sway_libs/src/string/) is an interface to implement dynamic length strings that are UTF-8 encoded.
 - [Signed Integers](./sway_libs/src/signed_integers/) is an interface to implement signed integers.
-- [Unsigned Fixed Point 64 bit](./sway_libs/src/fixed_point/ufp/ufp64) is an interface to implement fixed-point 64 bit numbers.
+- [Unsigned Fixed Point 64 bit](./sway_libs/src/fixed_point/ufp/) is an interface to implement fixed-point numbers.
```
Comment (bitzoic) [inline README.md:38]:
```suggestion
- [Unsigned Fixed Point Integers](./sway_libs/src/fixed_point/ufp/) is an interface to implement fixed-point numbers.
```

File: sway_libs/src/fixed_point/ufp/README.md [hunk 12]
```diff
@@ -0,0 +1,58 @@
+# Overview
+
+The Unsigned Fixed Point library provides a library to use fixed-point numbers in Sway. It has 2 distinct type: `UFP64` and `UFP128`. This type is stack allocated.
```
Comment (bitzoic) [inline sway_libs/src/fixed_point/ufp/README.md:3]:
```suggestion
The Unsigned Fixed Point Integer library provides a library to use fixed-point numbers in Sway. It has 2 distinct types: `UFP64` and `UFP128`. This type is stack allocated.

```

File: sway_libs/src/fixed_point/ufp/SPECIFICATION.md [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+
+Returns a zero value of type.
+
+### `from_uint()`
+
+Returns a fixed-point value equal to some unsigned value.
+
+### `floor()`
+
+Returns the biggest integer fixed-point value, that is smaller than the argument.
```
Comment (bitzoic) [inline sway_libs/src/fixed_point/ufp/SPECIFICATION.md:59]: Is this biggest or smallest? Same for `ceil()`

File: sway_libs/src/fixed_point/ufp/SPECIFICATION.md [hunk 14]
```diff
@@ -0,0 +1,86 @@
+# Overview
+
+This document provides an overview of the Unsigned Fixed Point library.
```
Comment (bitzoic) [inline sway_libs/src/fixed_point/ufp/SPECIFICATION.md:3]:
```suggestion
This document provides an overview of the Unsigned Fixed Point Integer library.

```

**Review by bitzoic — CHANGES_REQUESTED** [review]
Just docs comments

File: sway_libs/src/fixed_point/ufp/SPECIFICATION.md (hunk 13 above)
Comment (rostyslavtyshko) [reply sway_libs/src/fixed_point/ufp/SPECIFICATION.md:59]: Biggest integer that is smaller than the number i.e. 100.5 has 101 integer smaller than it, but `floor` has to return 100

Comment (bitzoic) [reply sway_libs/src/fixed_point/ufp/SPECIFICATION.md:59]: Nvm I think this is correct

File: sway_libs/src/fixed_point/ufp/README.md (hunk 12 above)
Comment (rostyslavtyshko) [reply sway_libs/src/fixed_point/ufp/README.md:3]: Fixed

File: README.md (hunk 11 above)
Comment (rostyslavtyshko) [reply README.md:38]: Fixed

**Review by bitzoic — APPROVED** [review]
Another one done 🥳

**Review by simonr0204 — APPROVED** [review]

**Pull request merged**